internal/rootcoord/root_coord.go

```diff
@@ -2465,3 +2466,197 @@ func segmentsOnlineReady(idxBuilt, segCount int) bool {
 	}
 	return false
 }
+
+// ExpireCredCache will call invalidate credential cache
+func (c *Core) ExpireCredCache(ctx context.Context, username string) {
+	req := proxypb.InvalidateCredCacheRequest{
+		Base: &commonpb.MsgBase{
+			MsgType:  0, //TODO, msg type
+			MsgID:    0, //TODO, msg id
+			SourceID: c.session.ServerID,
+		},
+		Username: username,
+	}
+	c.proxyClientManager.InvalidateCredentialCache(ctx, &req)
+}
+
+// UpdateCredCache will call update credential cache
+func (c *Core) UpdateCredCache(ctx context.Context, credInfo *etcdpb.CredentialInfo) {
+	req := proxypb.UpdateCredCacheRequest{
+		Base: &commonpb.MsgBase{
+			MsgType:  0, //TODO, msg type
+			MsgID:    0, //TODO, msg id
+			SourceID: c.session.ServerID,
+		},
+		Username: credInfo.Username,
+		Password: credInfo.EncryptedPassword,
+	}
+	c.proxyClientManager.UpdateCredentialCache(ctx, &req)
+}
+
+// ClearCredUsersCache will call clear credential usernames cache
+func (c *Core) ClearCredUsersCache(ctx context.Context) {
+	req := internalpb.ClearCredUsersCacheRequest{}
+	c.proxyClientManager.ClearCredUsersCache(ctx, &req)
+}
+
+// CreateCredential create new user and password
+// 	1. decode ciphertext password to raw password
+// 	2. encrypt raw password
+// 	3. save in to etcd
+func (c *Core) CreateCredential(ctx context.Context, in *milvuspb.CreateCredentialRequest) (*commonpb.Status, error) {
+	metrics.RootCoordCreateCredentialCounter.WithLabelValues(metrics.TotalLabel).Inc()
+	tr := timerecord.NewTimeRecorder("CreateCredential")
+	log.Debug("CreateCredential", zap.String("role", typeutil.RootCoordRole),
+		zap.String("username", in.Username))
+
+	if cred, _ := c.MetaTable.getCredential(in.Username); cred != nil {
+		return failStatus(commonpb.ErrorCode_UserExists, "user already exists:"+in.Username), nil
+	}
+	// update proxy's local cache
+	c.ClearCredUsersCache(ctx)
+	// insert to db
+	rawPassword, err := crypto.Base64Decode(in.Password)
+	if err != nil {
+		log.Error("decode password fail", zap.String("key", in.Username), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_UnexpectedError, "decode password fail key:"+in.Username), nil
+	}
+	encryptedPassword, err := crypto.PasswordEncrypt(rawPassword)
+	if err != nil {
+		log.Error("encrypt password fail", zap.String("key", in.Username), zap.Error(err))
+		return failStatus(commonpb.ErrorCode_UnexpectedError, "encrypt password fail key:"+in.Username), nil
+	}
+	credInfo := &etcdpb.CredentialInfo{
+		Username:          in.Username,
+		EncryptedPassword: encryptedPassword,
+	}
+	err = c.MetaTable.AddCredential(credInfo)
+	if err != nil {
+		log.Error("CreateCredential failed", zap.String("role", typeutil.RootCoordRole),
+			zap.String("username", in.Username), zap.Error(err))
+		metrics.RootCoordCreateCredentialCounter.WithLabelValues(metrics.FailLabel).Inc()
+		return failStatus(commonpb.ErrorCode_UnexpectedError, "CreateCredential failed: "+err.Error()), nil
```

Comment:
We'd better define a reasonable error code , it should not be UnexpectedError

Reply:
Ok